### PR TITLE
Feature registry eurka

### PR DESCRIPTION
**Pull request by sunrui849**
### Ⅰ. Describe what this PR did
merge to develop

### CODE REVIEW

File: common/src/main/java/com/alibaba/fescar/common/exception/EurekaRegistryException.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package com.alibaba.fescar.common.exception;
+
+
+ /**
+ * eureka registry exception
+ *
+ * @Author: rui_849217@163.com
```
Comment (slievrly) [inline common/src/main/java/com/alibaba/fescar/common/exception/EurekaRegistryException.java]: @author @date

File: discovery/pom.xml [hunk 2]
```diff
@@ -55,5 +55,20 @@
             <artifactId>testng</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.netflix.eureka</groupId>
```
Comment (slievrly) [inline discovery/pom.xml]: add parent pom

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package com.alibaba.fescar.discovery.registry;
+
+import com.netflix.appinfo.EurekaInstanceConfig;
+import com.netflix.appinfo.MyDataCenterInstanceConfig;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * The custom eureka instance config info
+ *
+ * @Author: rui_849217@163.com
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java]: @author @date copyright

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    private static final String EUREKA_CONFIG_SERVER_URL_KEY = "eureka.serviceUrl.default";
+    private static final String EUREKA_CONFIG_REFRESH_KEY = "eureka.client.refresh.interval";
+    private static final String EUREKA_CONFIG_METADATA_WEIGHT = "eureka.metadata.weight";
+    private static final int EUREKA_REFRESH_INTERVAL = 5;
+    private static final String DEFAULT_WEIGHT = "1";
+    private static final Configuration FILE_CONFIG = ConfigurationFactory.FILE_INSTANCE;
+
+    private static ApplicationInfoManager applicationInfoManager;
+    private static CustomEurekaInstanceConfig instanceConfig;
+    private static EurekaRegistryServiceImpl instance;
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: voliate？

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                } catch (Exception e) {
+                    close();
+                    throw new EurekaRegistryException("register eureka is error!",e);
+                }
+            }
+        }
+        return eurekaClient;
+    }
+
+    private void validAddress(InetSocketAddress address) {
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: NetUtils

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        applicationInfoManager.setInstanceStatus(InstanceInfo.InstanceStatus.UP);
+    }
+
+    @Override
+    public void unregister(InetSocketAddress address) throws Exception {
+        if (eurekaClient == null) {
+            return;
+        }
+        applicationInfoManager.setInstanceStatus(InstanceInfo.InstanceStatus.DOWN);
+        eurekaClient.shutdown();
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: if shutdown ，then register reInit？

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        assertEureka();
+        eurekaClient.unregisterEventListener(listener);
+    }
+
+    @Override
+    public List<InetSocketAddress> lookup(String key) throws Exception {
+        assertEureka();
+        List<InetSocketAddress> inetSocketAddressList = new ArrayList<>();
+
+        Application application = eurekaClient.getApplication(key.toUpperCase());
```
Comment (slievrly) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: first get clusterName from configCenter

File: common/src/main/java/com/alibaba/fescar/common/exception/EurekaRegistryException.java (hunk 1 above)
Comment (sunrui849) [reply common/src/main/java/com/alibaba/fescar/common/exception/EurekaRegistryException.java]: ok

File: discovery/pom.xml (hunk 2 above)
Comment (sunrui849) [reply discovery/pom.xml]: ok

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java (hunk 3 above)
Comment (sunrui849) [reply discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java]: ok

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java (hunk 7 above)
Comment (sunrui849) [reply discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: add return null

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java (hunk 5 above)
Comment (sunrui849) [reply discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: ok

File: discovery/pom.xml [hunk 8]
```diff
@@ -55,5 +55,17 @@
...
+        <dependency>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.netflix.archaius</groupId>
+            <artifactId>archaius-core</artifactId>
+        </dependency>
     </dependencies>
 </project>
```
Comment (jovany-wang) [inline discovery/pom.xml]: add a new line at the end of this file.

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public int getNonSecurePort() {
+        if (port == -1) {
+            return super.getNonSecurePort();
+        }
+        return port;
+    }
+
+    @Override
+    public String getAppname() {
```
Comment (jovany-wang) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java]: typo: `getAppName `
I think we should fix the parent class as well.

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    public void setIpAddress(String ipAddress) {
+        this.ipAddress = ipAddress;
+    }
+
+    public void setPort(int port) {
+        this.port = port;
+    }
+
+    public void setApplicationName(String applicationName) {
```
Comment (jovany-wang) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/CustomEurekaInstanceConfig.java]: make these `setApplicationName` and `getAppname ` names as pair:
setApplicationName and `getApplicationName`

or 

setAppName and `getAppName

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    @Override
+    public void unsubscribe(String cluster, EurekaEventListener listener) throws Exception {
+        assertEureka();
+        subscribeListener = false;
+        eurekaClient.unregisterEventListener(listener);
+    }
+
+    @Override
+    public List<InetSocketAddress> lookup(String key) throws Exception {
+        Configuration config = ConfigurationFactory.getInstance();
```
Comment (jovany-wang) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: I am considering that whether we need to make a `XXXConfig` class.

File: discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    private static void assertEureka(){
+        if (eurekaClient == null) {
+            throw new EurekaRegistryException("eureka client is not register,do this operation had to be register!");
+        }
+    }
+
+    private static String getInstanceId() {
+        return instanceConfig.getAppname() + "/" + instanceConfig.getIpAddress() + ":" + instanceConfig.getNonSecurePort();
```
Comment (jovany-wang) [inline discovery/src/main/java/com/alibaba/fescar/discovery/registry/EurekaRegistryServiceImpl.java]: It's better to use  String.format

**Comment by jovany-wang** [comment]
Note that you should rebase or merge current master to resolve conflicts.

**Pull request merged**